File: agents/ebpf/internal/otlpprofiles/convert.go
// Package otlpprofiles turns aggregated CPU samples into an OTLP profiles payload
// (docs/contracts/ebpf-profiler.md §5, docs/contracts/profiles.md).
//
// It is a second implementation of the shaping the Go agent does in its own internal/otlpprofiles, and that
// duplication is deliberate (D-148): agents/go is a published library, so it cannot depend on a shared
// package wired with a `replace` directive — the directive does not reach consumers, and `go get` would
// break for anyone importing the agent. The input differs too. The Go agent converts a pprof profile, which
// arrives as a dictionary of its own; here the input is already aggregated (stack, value) pairs from a BPF
// map, so there is no pprof to parse and no second sample type to split out.
//
// **Use the plain `go.opentelemetry.io/proto/otlp/...` modules, never the `slim` ones.** The slim modules
// replace the full ones rather than accompanying them: both register the same proto file, so importing one
// of each panics at init.
package otlpprofiles

import (
	"strconv"
	"strings"

	commonpb "go.opentelemetry.io/proto/otlp/common/v1"
	profilespb "go.opentelemetry.io/proto/otlp/profiles/v1development"
	resourcepb "go.opentelemetry.io/proto/otlp/resource/v1"
)

// MaxSamples bounds the samples of one profile. The receiving side caps what it will expand, so the agent
// must not send more than it would accept back (ebpf-profiler.md §6).
const MaxSamples = 100_000

// Frame is one call frame. Function is a resolved name where the binary carries symbols, and
// `<binary>+0x<offset>` where it does not — knowing which binary burned the CPU is still an answer, so an
// unresolved frame is reported rather than dropped (ebpf-profiler.md §7).
type Frame struct {
	Function string
	Address  uint64
}

// Sample is one aggregated call stack and the CPU nanoseconds measured on it.
//
// Frames are **leaf first**, the order OTLP carries. The receiving side decides which end a flame graph is
// drawn from, so nothing is reversed here.
type Sample struct {
	Frames []Frame
	Value  int64
}

// dict builds the payload's shared dictionary. Every table is interned by value, so a frame that appears in
// a thousand stacks costs one entry and a thousand int32s.
type dict struct {
	pb *profilespb.ProfilesDictionary

	strings   map[string]int32
	functions map[string]int32
	locations map[string]int32
	stacks    map[string]int32
}

func newDict() *dict {
	d := &dict{
		pb:        &profilespb.ProfilesDictionary{},
		strings:   map[string]int32{},
		functions: map[string]int32{},
		locations: map[string]int32{},
		stacks:    map[string]int32{},
	}
	d.str("") // index 0 is the empty string by convention
	return d
}

func (d *dict) str(s string) int32 {
	if i, ok := d.strings[s]; ok {
		return i
	}
	d.pb.StringTable = append(d.pb.StringTable, s)
	i := int32(len(d.pb.StringTable) - 1)
	d.strings[s] = i
	return i
}

func (d *dict) function(name string) int32 {
	if i, ok := d.functions[name]; ok {
		return i
	}
	d.pb.FunctionTable = append(d.pb.FunctionTable, &profilespb.Function{NameStrindex: d.str(name)})
	i := int32(len(d.pb.FunctionTable) - 1)
	d.functions[name] = i
	return i
}

// location interns one frame. Two frames with the same name at different addresses are different locations:
// the address is what tells two static functions of the same name apart.
func (d *dict) location(f Frame) int32 {
	key := f.Function + "\x00" + strconv.FormatUint(f.Address, 16)
	if i, ok := d.locations[key]; ok {
		return i
	}
	d.pb.LocationTable = append(d.pb.LocationTable, &profilespb.Location{
		Address: f.Address,
		Lines:   []*profilespb.Line{{FunctionIndex: d.function(f.Function)}},
	})
	i := int32(len(d.pb.LocationTable) - 1)
	d.locations[key] = i
	return i
}

func (d *dict) stack(frames []Frame) int32 {
	idx := make([]int32, 0, len(frames))
	var key strings.Builder
	for _, f := range frames {
		i := d.location(f)
		idx = append(idx, i)
		key.WriteString(strconv.Itoa(int(i)))
		key.WriteByte(',')
	}
	if i, ok := d.stacks[key.String()]; ok {
		return i
	}
	d.pb.StackTable = append(d.pb.StackTable, &profilespb.Stack{LocationIndices: idx})
	i := int32(len(d.pb.StackTable) - 1)
	d.stacks[key.String()] = i
	return i
}

// FromSamples converts one interval's samples into an OTLP profiles payload, or nil when there is nothing
// worth sending. An interval in which nothing ran produces no request at all (ebpf-profiler.md §5): an empty
// profile costs both sides a round trip and draws an empty flame graph.
//
// One profile, one unit. Only CPU nanoseconds are produced here, so unlike a pprof conversion there is no
// second sample type to split into a second profile — but the invariant it protects is the same, and the
// sample type is stated rather than assumed so a chart can say what its numbers mean.
func FromSamples(
	samples []Sample,
	resource []*commonpb.KeyValue,
	scopeName, scopeVersion string,
	startUnixNano uint64,
	durationNano uint64,
	periodNano int64,
) *profilespb.ProfilesData {
	d := newDict()
	out := make([]*profilespb.Sample, 0, min(len(samples), MaxSamples))
	for _, s := range samples {
		// A sample worth no time contributes nothing to any aggregate and would still cost a row
		// downstream. Negative is impossible from the kernel and is refused rather than trusted.
		if s.Value <= 0 || len(s.Frames) == 0 {
			continue
		}
		if len(out) >= MaxSamples {
			break
		}
		out = append(out, &profilespb.Sample{StackIndex: d.stack(s.Frames), Values: []int64{s.Value}})
	}
	if len(out) == 0 {
		return nil
	}

	cpu, nanos := d.str("cpu"), d.str("nanoseconds")
	return &profilespb.ProfilesData{
		Dictionary: d.pb,
		ResourceProfiles: []*profilespb.ResourceProfiles{{
			Resource: &resourcepb.Resource{Attributes: resource},
			ScopeProfiles: []*profilespb.ScopeProfiles{{
				Scope: &commonpb.InstrumentationScope{Name: scopeName, Version: scopeVersion},
				Profiles: []*profilespb.Profile{{
					SampleType:   &profilespb.ValueType{TypeStrindex: cpu, UnitStrindex: nanos},
					PeriodType:   &profilespb.ValueType{TypeStrindex: cpu, UnitStrindex: nanos},
					Period:       periodNano,
					Samples:      out,
					TimeUnixNano: startUnixNano,
					DurationNano: durationNano,
				}},
			}},
		}},
	}
}

File: agents/node/src/profiles/convert.ts
// V8 CPU profile -> OTLP profiles (docs/contracts/profiles.md).
//
// Written by hand, as OTLP/JSON, for two reasons. The OpenTelemetry JS SDK has no profiles exporter —
// profiles are not part of its stable surface — so there is nothing to reuse; and the ingest accepts JSON
// and re-encodes it to protobuf itself, so the agent needs no protobuf runtime to speak the wire format.
//
// The field names below are the proto's own (snake_case), which protojson accepts, and int64 fields are
// strings: a JSON number cannot carry them without losing digits past 2^53, and a duration in nanoseconds
// passes that in under two hours.
//
// **Frames are function names only.** The server's stack model is a list of names (internal/profiles
// resolveStack reads FunctionTable.name_strindex), so file and line are deliberately not encoded: they
// would be interned, sent, and then dropped.

/** The shape `inspector`'s Profiler.takeProfile returns (Profiler.Profile). */
export interface V8Profile {
  nodes: V8Node[];
  /** Microseconds since an arbitrary epoch. */
  startTime: number;
  endTime: number;
  /** Sampled node ids, and the microseconds between consecutive samples. */
  samples?: number[];
  timeDeltas?: number[];
}

export interface V8Node {
  id: number;
  callFrame: { functionName?: string; url?: string; lineNumber?: number; columnNumber?: number };
  hitCount?: number;
  children?: number[];
}

export interface KeyValue {
  key: string;
  value: { stringValue: string };
}

/** Caps, mirroring the Go agent so one producer cannot be an order of magnitude heavier than another. */
export const MAX_SAMPLES = 20_000;
export const MAX_FRAME_BYTES = 512;

interface Dict {
  string_table: string[];
  function_table: { name_strindex: number }[];
  location_table: { lines: { function_index: number }[] }[];
  stack_table: { location_indices: number[] }[];
}

class Builder {
  readonly dict: Dict = { string_table: [], function_table: [], location_table: [], stack_table: [] };
  private readonly strings = new Map<string, number>();
  private readonly functions = new Map<string, number>();
  private readonly locations = new Map<number, number>();
  private readonly stacks = new Map<string, number>();

  str(s: string): number {
    const found = this.strings.get(s);
    if (found !== undefined) return found;
    const i = this.dict.string_table.length;
    this.dict.string_table.push(s);
    this.strings.set(s, i);
    return i;
  }

  /** One location per V8 node, one function per name: the dictionary is interned, not repeated. */
  location(nodeId: number, name: string): number {
    const found = this.locations.get(nodeId);
    if (found !== undefined) return found;
    let fn = this.functions.get(name);
    if (fn === undefined) {
      fn = this.dict.function_table.length;
      this.dict.function_table.push({ name_strindex: this.str(name.slice(0, MAX_FRAME_BYTES)) });
      this.functions.set(name, fn);
    }
    const i = this.dict.location_table.length;
    this.dict.location_table.push({ lines: [{ function_index: fn }] });
    this.locations.set(nodeId, i);
    return i;
  }

  /** `indices` must already be leaf first: OTLP orders a stack that way, and the server reverses it. */
  stack(indices: number[]): number {
    const key = indices.join(',');
    const found = this.stacks.get(key);
    if (found !== undefined) return found;
    const i = this.dict.stack_table.length;
    this.dict.stack_table.push({ location_indices: indices });
    this.stacks.set(key, i);
    return i;
  }
}

/** The name a frame is stored under. V8 leaves the main script and native frames unnamed. */
function frameName(n: V8Node): string {
  const fn = n.callFrame.functionName;
  return fn && fn.length > 0 ? fn : '(anonymous)';
}

/**
 * Converts one V8 CPU profile into an OTLP ProfilesData object, or null when the process was idle.
 *
 * Null rather than an empty payload on purpose: the ingest accepts a profile-less body and produces
 * nothing, so sending one would be a request that costs both sides and says nothing.
 */
export function fromV8Profile(
  p: V8Profile,
  resource: KeyValue[],
  scope: { name: string; version: string },
  endWallMs: number,
): Record<string, unknown> | null {
  if (!p || !Array.isArray(p.nodes) || p.nodes.length === 0) return null;
  const samples = p.samples ?? [];
  const deltas = p.timeDeltas ?? [];
  if (samples.length === 0) return null;

  const byId = new Map<number, V8Node>();
  for (const n of p.nodes) byId.set(n.id, n);
  // V8 gives children; a stack needs parents.
  const parent = new Map<number, number>();
  for (const n of p.nodes) for (const c of n.children ?? []) parent.set(c, n.id);

  // Nanoseconds per sampled node: a sample's delta is the time since the previous one, in microseconds.
  const nanos = new Map<number, number>();
  for (let i = 0; i < samples.length; i++) {
    const d = deltas[i];
    // The only place a zero is dropped: a sample worth no time contributes nothing to any aggregate and
    // would still cost a row downstream. Everything in `nanos` below is therefore strictly positive.
    if (typeof d !== 'number' || d <= 0) continue;
    nanos.set(samples[i], (nanos.get(samples[i]) ?? 0) + d * 1000);
  }

  const b = new Builder();
  const out: { stack_index: number; values: string[] }[] = [];
  for (const [nodeId, value] of nanos) {
    if (out.length >= MAX_SAMPLES) break;
    const indices: number[] = [];
    for (let id: number | undefined = nodeId; id !== undefined; id = parent.get(id)) {
      const n = byId.get(id);
      if (!n) break;
      indices.push(b.location(n.id, frameName(n))); // leaf first, walking up to the root
    }
    if (indices.length === 0) continue;
    out.push({ stack_index: b.stack(indices), values: [String(value)] });
  }
  if (out.length === 0) return null;

  const durationNs = Math.max(0, Math.round((p.endTime - p.startTime) * 1000));
  return {
    dictionary: b.dict,
    resource_profiles: [
      {
        resource: { attributes: resource },
        scope_profiles: [
          {
            scope: { name: scope.name, version: scope.version },
            profiles: [
              {
                // The server refuses a profile whose sample type resolves to an empty string: a chart of
                // unnamed numbers is worse than no chart.
                sample_type: { type_strindex: b.str('cpu'), unit_strindex: b.str('nanoseconds') },
                samples: out,
                time_unix_nano: String(Math.round(endWallMs) * 1_000_000),
                duration_nano: String(durationNs),
              },
            ],
          },
        ],
      },
    ],
  };
}
